include/onnxruntime/core/session/onnxruntime_c_api.h

```diff
@@ -1174,6 +1174,21 @@ struct OrtApi {
    * Get the current device id of the GPU execution provider (cuda/tensorrt/rocm).
    */
   ORT_API2_STATUS(GetCurrentGpuDeviceId, _In_ int* device_id);
+
+  /**
+  * Only works on windows system for release(1.7.3)
+  * Set thread affinity for intra thread pool threads.
+  * affinity_string is of format:
+  * "group,processor_mask;group,processor_mask;group,processor_mask;..."
+  * "group" is an ordinal number of a processor group,
+  * "processor_mask" is a bitmask of processors that a thread is expected to attach to.
+  * e.g., processor 1 and 2 in group 1 will be represented as:
+  * 0,3
+  * where processor 2 and 4 in group 2 will be set to:
+  * 1,10
+  * the number of "group,processor_mask" pair should be intra_op_num_threads - 1, since ort will skip setting affinity for the main thread
+  */
+  ORT_API2_STATUS(SetGlobalIntraOpThreadAffinity, _Inout_ OrtThreadingOptions* tp_options, const char* affinity_string);
```

Comment:
I suggest an array of integers

Reply:
@yuslepukhin Yeah, initially this was implemented as a quick fix for unblocking Bing Ads to expedite their migration to NUMA machines. It's a bit unfortunate that we've to patch a release as old as 1.7. We can think of a better way to express these affinities in the API. A 1-d array of integers won't work since we need to encode the group number as well. Suggestions?

Reply:
I would still suggest an array of integers, plus a separate array of the group lengths. This approach is already familiar to users as we employ it with some strings API.

Reply:
> I would still suggest an array of integers, plus a separate array of the group lengths. This approach is already familiar to users as we employ it with some strings API.

Meaning we got to expose two APIs, one as array of integers representing processors and another for an array of groups?

Reply:
Meaning we have one API with more arguments.

Reply:
> I suggest an array of integers. So we are going to be stuck with the API that works for a specific release? We should add an API that can be used everywhere. People, are going to write code against it.

Thanks for the thinking! :-)
So with an array of integers how we tell which one is for group and which one for processor?
Assume a simple case, that with group of integers we have like:

0,1,0,2,1,1,1,1,3

meaning we have 4 thread affinity settings for 4 threads:

0,1
1,2
The first thread will run in group 0 processor 1, the second thread run in group 1 processor 2 ...
This brings a few problems - among which the biggest one is what if there are 64 processor each group and customer want the first thread attach to 32 processors in group 0, and the second thread to 32 processors in group 1? Are we looking at:

0,1
0,2
0,3
...
0,32
1,1
1.2
...
1,32

?

Reply:
However, the issue here also is this look very Windows specific. On Linux there is not really any groups, just the number of logical processors.
I think any API we publish, should be OK for different platforms.

For Linux the first array would look like this (thread 0 attaches to procs `(0, 1)`, thread 1 to `(2,3)`
`0,1,2,3`
Second array is simply `(2,2)`

We need to reconcile this with Windows.

Reply:
Exactly - things be very different among platforms. That's we are just using a string there which could take varied forms accordingly.

Reply:
We should make them not different. The user cannot write different code for different platforms using our API. You have a choice of
- make it right now
- make it platform dependent and then it will be more difficult to make it right.

Reply:
The API is platform neutral, we are giving there be a affinity string which could be of desired format on varied platforms.

Reply:
> * make it platform dependent and then it will be more difficult to make it right.

I don't get this part - where are we having this "platform dependent" API?

Reply:
Formats of the string is now generic, pls see latest API usage, thanks!